### PR TITLE
Rewrite status override logic in registry using a rule framework

### PR DESCRIPTION
**Pull request by drtechniko**

### CODE REVIEW

**Comment by drtechniko** [comment]
@qiangdavidliu Here you go. Let's chat so I can fill you in if you have any questions.

**Comment by drtechniko** [comment]
There is more room for making this more DI friendly (AwsInstanceRegistry for example) but I didn't want to stir the waters too much in this PR.

File: eureka-core/src/main/java/com/netflix/eureka/registry/rule/StatusOverrideResult.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+public class StatusOverrideResult {
+
+    public static StatusOverrideResult NO_MATCH = new StatusOverrideResult(false, null);
+
+    public static StatusOverrideResult matchingStatus(InstanceInfo.InstanceStatus status) {
+        return new StatusOverrideResult(true, status);
+    }
+
+    // Does the rule match?
+    private boolean matches;
```
Comment (qiangdavidliu) [inline eureka-core/src/main/java/com/netflix/eureka/registry/rule/StatusOverrideResult.java]: Make these final.

File: eureka-core/src/main/java/com/netflix/eureka/registry/rule/DefaultStatusRule.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This is the default rule. It should be at the bottom of each rule chain.
+ * It matches always and returns the current status of the instance.
+ *
+ * Created by Nikos Michalakis on 7/13/16.
+ */
+public class DefaultStatusRule implements InstanceStatusOverrideRule {
```
Comment (qiangdavidliu) [inline eureka-core/src/main/java/com/netflix/eureka/registry/rule/DefaultStatusRule.java]: Rename to something like AlwaysMatchStatusRule to better describe what this rule does?

File: eureka-core/src/main/java/com/netflix/eureka/registry/rule/FirstMatchWinsCompositeRule.java [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+/**
+ * This rule takes an ordered list of rules and returns the result of the first match or the
+ * result of the {@link DefaultStatusRule}.
+ *
+ * Created by Nikos Michalakis on 7/13/16.
+ */
+public class FirstMatchWinsCompositeRule implements InstanceStatusOverrideRule {
+
+    private InstanceStatusOverrideRule[] rules;
```
Comment (qiangdavidliu) [inline eureka-core/src/main/java/com/netflix/eureka/registry/rule/FirstMatchWinsCompositeRule.java]: These can all be final

File: eureka-core/src/main/java/com/netflix/eureka/registry/rule/DownOrStartingRule.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
+package com.netflix.eureka.registry.rule;
+
+import com.netflix.appinfo.InstanceInfo;
+import com.netflix.eureka.lease.Lease;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * If the instance is STARTING or DOWN then we return that.
+ * This rule has precedence over the {@link OverrideExistsRule}.
```
Comment (qiangdavidliu) [inline eureka-core/src/main/java/com/netflix/eureka/registry/rule/DownOrStartingRule.java]: This comment is misleading. From the rule interface design, each individual rule should not have understanding of where it applies in relation to other rules, and it is up to the definer at the top level to define them in the appropriate hierarchy. Perhaps move this comment to the *registry.java's?

File: eureka-core/src/main/java/com/netflix/eureka/registry/rule/AsgEnabledRule.java [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+import com.netflix.eureka.aws.AwsAsgUtil;
+import com.netflix.eureka.lease.Lease;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This is a rule that checks if the ASG for an instance is enabled or not and if not then it brings the instance
+ * OUT_OF_SERVICE.
+ *
+ * This rule should be placed after the {@link OverrideExistsRule} and before the {@link LeaseExistsRule}.
```
Comment (qiangdavidliu) [inline eureka-core/src/main/java/com/netflix/eureka/registry/rule/AsgEnabledRule.java]: Rules themselves should Not understand their hierarchy.

File: eureka-core/src/test/java/com/netflix/eureka/registry/AwsInstanceRegistryTest.java [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+
+    @Test
+    public void testOverridesWithAsgEnabledThenDisabled() {
+        // Regular registration first
+        InstanceInfo myInstance = createLocalUpInstanceWithAsg(LOCAL_REGION_INSTANCE_1_HOSTNAME);
+        registerInstanceLocally(myInstance);
+        verifyLocalInstanceStatus(myInstance.getId(), InstanceStatus.UP);
+
+        // Now we disable the ASG and we should expect OUT_OF_SERVICE status.
+        ((AwsInstanceRegistry) registry).getAwsAsgUtil().setStatus("ASG-YO-HO", false);
```
Comment (qiangdavidliu) [inline eureka-core/src/test/java/com/netflix/eureka/registry/AwsInstanceRegistryTest.java]: Why not use myInstance.getASGName() here? That way there won't be magic strings.

File: eureka-core/src/test/java/com/netflix/eureka/registry/AwsInstanceRegistryTest.java [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+
+        myInstance = createLocalUpInstanceWithAsg(LOCAL_REGION_INSTANCE_1_HOSTNAME);
+        registerInstanceLocally(myInstance);
+        verifyLocalInstanceStatus(myInstance.getId(), InstanceStatus.OUT_OF_SERVICE);
+
+        // Now we re-enable the ASG and we should expect UP status.
+        ((AwsInstanceRegistry) registry).getAwsAsgUtil().setStatus("ASG-YO-HO", true);
```
Comment (qiangdavidliu) [inline eureka-core/src/test/java/com/netflix/eureka/registry/AwsInstanceRegistryTest.java]: Why not use myInstance.getASGName() here? That way there won't be magic strings.

**Comment by qiangdavidliu** [comment]
Minor quibbles, but LGTM. Thanks for the PR.

File: eureka-core/src/main/java/com/netflix/eureka/registry/rule/DownOrStartingRule.java (hunk 4 above)
Comment (drtechniko) [inline eureka-core/src/main/java/com/netflix/eureka/registry/rule/DownOrStartingRule.java]: good point. I updated all the rule comments to be self-contained. Moved descriptions of rule order to the registries.

File: eureka-core/src/test/java/com/netflix/eureka/registry/AwsInstanceRegistryTest.java (hunk 6 above)
Comment (drtechniko) [inline eureka-core/src/test/java/com/netflix/eureka/registry/AwsInstanceRegistryTest.java]: good idea. done.

File: eureka-core/src/main/java/com/netflix/eureka/registry/rule/StatusOverrideResult.java (hunk 1 above)
Comment (drtechniko) [inline eureka-core/src/main/java/com/netflix/eureka/registry/rule/StatusOverrideResult.java]: done

File: eureka-core/src/main/java/com/netflix/eureka/registry/rule/DefaultStatusRule.java (hunk 2 above)
Comment (drtechniko) [inline eureka-core/src/main/java/com/netflix/eureka/registry/rule/DefaultStatusRule.java]: renamed to AlwaysMatchInstanceStatusRule

File: eureka-core/src/main/java/com/netflix/eureka/registry/rule/FirstMatchWinsCompositeRule.java (hunk 3 above)
Comment (drtechniko) [inline eureka-core/src/main/java/com/netflix/eureka/registry/rule/FirstMatchWinsCompositeRule.java]: done

File: eureka-core/src/main/java/com/netflix/eureka/registry/rule/AsgEnabledRule.java (hunk 5 above)
Comment (drtechniko) [inline eureka-core/src/main/java/com/netflix/eureka/registry/rule/AsgEnabledRule.java]: done

**Pull request merged**